File: clients-api/src/shared/outbox/outbox.document.ts
export const OUTBOX_COLLECTION = 'outbox';

export enum OutboxStatus {
  PENDING = 'PENDING',
  IN_FLIGHT = 'IN_FLIGHT',
  PUBLISHED = 'PUBLISHED',
}

export const UNPUBLISHED_STATUSES: readonly OutboxStatus[] = Object.freeze([
  OutboxStatus.PENDING,
  OutboxStatus.IN_FLIGHT,
]);

export interface OutboxMessage {
  readonly id: string;
  readonly topic: string;
  readonly key: string;
  readonly version: number;
  readonly payload: object;
}

export interface OutboxDocument {
  readonly _id: string;
  readonly topic: string;
  readonly key: string;
  readonly version: number;
  readonly payload: object;
  readonly status: OutboxStatus;
  readonly attempts: number;
  readonly createdAt: Date;
  readonly availableAt: Date;
  readonly leaseOwner?: string;
  readonly leaseUntil?: Date;
  readonly publishedAt?: Date;
  readonly lastError?: string;
}

export function pendingOutboxDocument(message: OutboxMessage, createdAt: Date): OutboxDocument {
  return {
    _id: message.id,
    topic: message.topic,
    key: message.key,
    version: message.version,
    payload: message.payload,
    status: OutboxStatus.PENDING,
    attempts: 0,
    createdAt,
    availableAt: createdAt,
  };
}

export function toOutboxMessage(document: OutboxDocument): OutboxMessage {
  return {
    id: document._id,
    topic: document.topic,
    key: document.key,
    version: document.version,
    payload: document.payload,
  };
}

File: order-processor/src/main/java/com/grupomariposa/orders/infrastructure/persistence/document/OutboxDocument.java
package com.grupomariposa.orders.infrastructure.persistence.document;

import java.time.Instant;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

@Document(collection = OutboxDocument.COLLECTION)
public record OutboxDocument(
        @Id String id,
        String orderId,
        int eventVersion,
        String topic,
        String key,
        String payload,
        OutboxStatus status,
        int attempts,
        Instant leaseUntil,
        String leaseOwner,
        Instant availableAt,
        Instant createdAt,
        Instant publishedAt) {

    public static final String COLLECTION = "outbox";
}
